From 06b1982554afbee4bef105cc3a2415a1f69d3fa6 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Tue, 11 Aug 2026 13:57:57 -0400
Subject: [PATCH 269/273] win32u: Implement common EDID emulation code.

Based on patches by Paul Gofman and Jactry Zeng.
---
 dlls/win32u/sysparams.c   | 230 ++++++++++++++++++++++++++++++++++----
 include/wine/gdi_driver.h |  34 ++++++
 2 files changed, 240 insertions(+), 24 deletions(-)

diff --git a/dlls/win32u/sysparams.c b/dlls/win32u/sysparams.c
index b2f60d833f5..8b523e5921a 100644
--- a/dlls/win32u/sysparams.c
+++ b/dlls/win32u/sysparams.c
@@ -130,21 +130,6 @@ struct source
     DEVMODEW *modes;
 };
 
-#define MONITOR_INFO_HAS_MONITOR_ID 0x00000001
-#define MONITOR_INFO_HAS_MONITOR_NAME 0x00000002
-#define MONITOR_INFO_HAS_PREFERRED_MODE 0x00000004
-struct edid_monitor_info
-{
-    unsigned int flags;
-    /* MONITOR_INFO_HAS_MONITOR_ID */
-    unsigned short manufacturer, product_code;
-    char monitor_id_string[8];
-    /* MONITOR_INFO_HAS_MONITOR_NAME */
-    WCHAR monitor_name[14];
-    /* MONITOR_INFO_HAS_PREFERRED_MODE */
-    unsigned int preferred_width, preferred_height;
-};
-
 struct monitor
 {
     LONG refcount;
@@ -2096,6 +2081,191 @@ static BOOL write_monitor_to_registry( struct monitor *monitor, const BYTE *edid
     return TRUE;
 }
 
+static BYTE edid_checksum( BYTE *data )
+{
+    UINT i;
+    BYTE ret = 0;
+
+    for (i = 0; i < 127; i++) ret += data[i];
+
+    return 0x100 - ret;
+}
+
+static BYTE edid_max_luminance(float nits)
+{
+    if (nits == 0.0f) return 0;
+    return ceilf((logf(nits / 50.0f) / logf(2.0f)) * 32.0f);
+}
+
+static void edid_detailed_timing_desc( const struct edid_monitor_info *info, BYTE *data )
+{
+    double refresh = info->preferred_refresh <= 0.0 ? 60.0 : info->preferred_refresh;
+    const unsigned h_front_porch = 8, h_sync_pulse = 32, h_back_porch = 40;
+    const unsigned v_front_porch = 6, v_sync_pulse = 8, v_back_porch = 40;
+    unsigned h_blanking, v_blanking, h_total, v_total, pixel_clock;
+
+    h_blanking = h_front_porch + h_sync_pulse + h_back_porch;
+    v_blanking = v_front_porch + v_sync_pulse + v_back_porch;
+    h_total = info->preferred_width + h_blanking;
+    v_total = info->preferred_height + v_blanking;
+    pixel_clock = round(h_total * v_total * refresh / 1e4);
+
+    data[0] = pixel_clock & 0xff;
+    data[1] = (pixel_clock >> 8) & 0xff;
+    data[2] = info->preferred_width;
+    data[3] = h_total - info->preferred_width;
+    data[4] = (((h_total - info->preferred_width) >> 8) & 0xf);
+    data[4] |= (((info->preferred_width >> 8) & 0xf) << 4);
+    data[5] = info->preferred_height;
+    data[6] = v_total - info->preferred_height;
+    data[7] = (((v_total - info->preferred_height) >> 8) & 0xf);
+    data[7] |= (((info->preferred_height >> 8) & 0xf) << 4);
+    data[8] = h_front_porch;
+    data[9] = h_sync_pulse;
+    data[10] = ((v_front_porch & 0xf) << 4) | (v_sync_pulse & 0xf);
+    data[11] = (((h_front_porch >> 8) & 3) << 6) | (((h_sync_pulse >> 8) & 3) << 4);
+    data[11] |= (((v_front_porch >> 4) & 3) << 2) | ((v_sync_pulse >> 4) & 3);
+    data[12] = info->width_mm;
+    data[13] = info->height_mm;
+    data[14] = (((info->width_mm >> 8) & 0xf) << 4) | ((info->height_mm >> 8) & 0xf);
+    data[17] = 0x1e;
+}
+
+static BOOL get_edid_from_monitor_info( const struct edid_monitor_info *info, BYTE **edid_out, UINT *edid_len )
+{
+    UINT extensions = 0, i;
+    BYTE *edid, *p;
+    char model[14] = {0};
+
+    if (!(info->flags & MONITOR_INFO_HAS_PREFERRED_MODE)) return FALSE;
+    if (!(info->flags & MONITOR_INFO_HAS_PRIMARIES)) return FALSE;
+    if (!(info->flags & MONITOR_INFO_HAS_PHYSICAL_DIMENSIONS)) return FALSE;
+
+    if (info->flags & MONITOR_INFO_HAS_CTA861_EXT) extensions++;
+
+    *edid_len = 128 + extensions * 128;
+    if (!(*edid_out = edid = calloc( *edid_len, sizeof(BYTE) )))
+    {
+        *edid_len = 0;
+        return FALSE;
+    }
+
+    *(ULONG64*)edid = 0x00ffffffffffff00;
+
+    if (info->flags & MONITOR_INFO_HAS_MONITOR_ID)
+    {
+        edid[8] = info->manufacturer & 0xff;
+        edid[9] = (info->manufacturer >> 8) & 0xff;
+        edid[10] = info->product_code & 0xff;
+        edid[11] = (info->product_code >> 8) & 0xff;
+    }
+
+    if (info->flags & MONITOR_INFO_HAS_SERIAL_NUMBER)
+        *(unsigned int*)(edid + 12) = info->serial_number;
+
+    edid[16] = 0xff;
+    edid[17] = 31; /* 2021 */
+    edid[18] = 1;
+    edid[19] = 4;
+    edid[20] = 0xf5; /* digital input, reserved bpc, display port */
+    edid[21] = round( info->width_mm / 10.0 );
+    edid[22] = round( info->height_mm / 10.0 );
+    edid[23] = 0x78; /* 2.2 gamma */
+    edid[24] = info->srgb ? 0x6 : 0x2;
+
+    if (info->srgb)
+    {
+        edid[25] = 0xee;
+        edid[26] = 0x91;
+        edid[27] = 0xa3;
+        edid[28] = 0x54;
+        edid[29] = 0x4c;
+        edid[30] = 0x99;
+        edid[31] = 0x26;
+        edid[32] = 0x0f;
+        edid[33] = 0x50;
+        edid[34] = 0x54;
+    }
+    else
+    {
+        edid[25] = ((info->r_x & 0x3) << 6) | ((info->r_y & 0x3) << 4) |
+                    ((info->g_x & 0x3) << 2) | (info->g_y & 0x3);
+        edid[26] = ((info->b_x & 0x3) << 6) | ((info->b_y & 0x3) << 4) |
+                    ((info->w_x & 0x3) << 2) | (info->w_y & 0x3);
+        edid[27] = (info->r_x & 0x3fc) >> 2;
+        edid[28] = (info->r_y & 0x3fc) >> 2;
+        edid[29] = (info->g_x & 0x3fc) >> 2;
+        edid[30] = (info->g_y & 0x3fc) >> 2;
+        edid[31] = (info->b_x & 0x3fc) >> 2;
+        edid[32] = (info->b_y & 0x3fc) >> 2;
+        edid[33] = (info->w_x & 0x3fc) >> 2;
+        edid[34] = (info->w_y & 0x3fc) >> 2;
+    }
+
+    for (i = 0; i < 16; i++) edid[38 + i] = 1;
+
+    p = edid + 54;
+    edid_detailed_timing_desc(info, p);
+
+    p += 18;
+    p[3] = 0xfc;
+
+    if (info->flags & MONITOR_INFO_HAS_MONITOR_NAME)
+        unicodez_to_ascii( model, info->monitor_name );
+    else
+        strcpy( model, "Wine Monitor" );
+
+    /* terminate the string with \n */
+    for (i = 0; i < sizeof(model)-1; i++)
+    {
+        if (!model[i])
+        {
+            model[i++] = '\n';
+            break;
+        }
+    }
+
+    /* then spaces after the \n */
+    for (; i < sizeof(model)-1; i++)
+        if (!model[i]) model[i] = ' ';
+
+    memcpy( (char *)p + 5, model, sizeof(model)-1 );
+
+    /* TODO: Add a way for drivers to use the remaining descriptors */
+    p += 18;
+    p[3] = 0x10;
+    p += 18;
+    p[3] = 0x10;
+
+    edid[126] = extensions;
+    edid[127] = edid_checksum( edid );
+
+    p = edid;
+
+    if (info->flags & MONITOR_INFO_HAS_CTA861_EXT)
+    {
+        p += 128;
+
+        p[0] = 2;
+        p[1] = 3;
+        p[2] = 0xb;
+
+        p[4] = (0x7 << 5) | 0x6;
+        p[5] = 6;
+
+        /* HDR static metadata block */
+        p[6] = 0x7; /* ST 2084 | SDR | HDR */
+        p[7] = 1;
+        p[8] = edid_max_luminance( info->max_cll );
+        p[9] = edid_max_luminance( info->max_fall );
+        p[10] = 0; /* many apps implement min luminance incorrectly */
+
+        p[127] = edid_checksum( p );
+    }
+
+    return TRUE;
+}
+
 static void add_monitor( const struct gdi_monitor *gdi_monitor, void *param )
 {
     struct device_manager_ctx *ctx = param;
@@ -2103,6 +2273,8 @@ static void add_monitor( const struct gdi_monitor *gdi_monitor, void *param )
     struct source *source;
     char buffer[MAX_PATH];
     char monitor_id_string[16];
+    BYTE *edid, generated_edid = 0;
+    UINT edid_len;
 
     assert( !list_empty( &sources ) );
     source = LIST_ENTRY( list_tail( &sources ), struct source, entry );
@@ -2117,7 +2289,16 @@ static void add_monitor( const struct gdi_monitor *gdi_monitor, void *param )
 
     TRACE( "%u %s %s\n", monitor->id, wine_dbgstr_rect(&gdi_monitor->rc_monitor), wine_dbgstr_rect(&gdi_monitor->rc_work) );
 
-    get_monitor_info_from_edid( &monitor->edid_info, gdi_monitor->edid, gdi_monitor->edid_len );
+    edid = gdi_monitor->edid;
+    edid_len = gdi_monitor->edid_len;
+
+    if (edid) get_monitor_info_from_edid( &monitor->edid_info, edid, edid_len );
+    else
+    {
+        generated_edid = get_edid_from_monitor_info( &gdi_monitor->edid_info, &edid, &edid_len );
+        monitor->edid_info = gdi_monitor->edid_info;
+    }
+
     if (monitor->edid_info.flags & MONITOR_INFO_HAS_MONITOR_ID)
         strcpy( monitor_id_string, monitor->edid_info.monitor_id_string );
     else
@@ -2127,18 +2308,19 @@ static void add_monitor( const struct gdi_monitor *gdi_monitor, void *param )
     snprintf( monitor->path, sizeof(monitor->path), "DISPLAY\\%s\\%04X&%04X", monitor_id_string, source->id, monitor->id );
     set_reg_ascii_value( source->key, buffer, monitor->path );
 
-    if (!write_monitor_to_registry( monitor, gdi_monitor->edid, gdi_monitor->edid_len ))
+    if (!write_monitor_to_registry( monitor, edid, edid_len ))
     {
         WARN( "Failed to write monitor %p to registry\n", monitor );
+        if (generated_edid) free( edid );
         monitor_release( monitor );
+        return;
     }
-    else
-    {
-        list_add_tail( &monitors, &monitor->entry );
-        TRACE( "created monitor %p for source %p\n", monitor, source );
-        source->monitor_count++;
-        ctx->monitor_count++;
-    }
+
+    if (generated_edid) free( edid );
+    list_add_tail( &monitors, &monitor->entry );
+    TRACE( "created monitor %p for source %p\n", monitor, source );
+    source->monitor_count++;
+    ctx->monitor_count++;
 }
 
 static UINT add_screen_size( SIZE *sizes, UINT count, SIZE size )
diff --git a/include/wine/gdi_driver.h b/include/wine/gdi_driver.h
index d52f6ec3734..d6bf25821e3 100644
--- a/include/wine/gdi_driver.h
+++ b/include/wine/gdi_driver.h
@@ -339,12 +339,46 @@ struct pci_id
     UINT16 revision;
 };
 
+#define MONITOR_INFO_HAS_MONITOR_ID          0x00000001
+#define MONITOR_INFO_HAS_MONITOR_NAME        0x00000002
+#define MONITOR_INFO_HAS_PREFERRED_MODE      0x00000004
+#define MONITOR_INFO_HAS_PHYSICAL_DIMENSIONS 0x00000008
+#define MONITOR_INFO_HAS_SERIAL_NUMBER       0x00000010
+#define MONITOR_INFO_HAS_PRIMARIES           0x00000020
+#define MONITOR_INFO_HAS_CTA861_EXT          0x00000040
+
+struct edid_monitor_info
+{
+    unsigned int flags;
+    /* MONITOR_INFO_HAS_MONITOR_ID */
+    unsigned short manufacturer, product_code;
+    char monitor_id_string[8];
+    /* MONITOR_INFO_HAS_MONITOR_NAME */
+    WCHAR monitor_name[14];
+    /* MONITOR_INFO_HAS_PREFERRED_MODE */
+    unsigned int preferred_width, preferred_height;
+    double preferred_refresh;
+    /* MONITOR_INFO_HAS_SERIAL_NUMBER */
+    unsigned int serial_number;
+    /* MONITOR_INFO_HAS_PHYSICAL_DIMENSIONS */
+    unsigned int width_mm, height_mm;
+    /* MONITOR_INFO_HAS_PRIMARIES */
+    BOOL srgb;
+    unsigned int r_x, r_y;
+    unsigned int g_x, g_y;
+    unsigned int b_x, b_y;
+    unsigned int w_x, w_y;
+    /* MONITOR_INFO_HAS_CTA861_EXT */
+    float max_cll, max_fall;
+};
+
 struct gdi_monitor
 {
     RECT rc_monitor;      /* RcMonitor in MONITORINFO struct */
     RECT rc_work;         /* RcWork in MONITORINFO struct */
     unsigned char *edid;  /* Extended Device Identification Data */
     UINT edid_len;
+    struct edid_monitor_info edid_info; /* EDID info to generate an EDID */
     BOOL hdr_enabled;
 };
 
-- 
2.55.0

